homeassistant/components/fritz/switch.py

```diff
@@ -0,0 +1,639 @@
+    """Get list of deflection entities."""
+
+    _LOGGER.debug("Setting up %s switches", SWITCH_TYPE_DEFLECTION)
+
+    service_name = "X_AVM-DE_OnTel"
+    deflections_response = service_call_action(
+        fritzbox_tools, service_name, "1", "GetNumberOfDeflections"
+    )
+    if not deflections_response:
+        _LOGGER.debug("The FRITZ!Box has no %s options", SWITCH_TYPE_DEFLECTION)
+        return []
+
+    _LOGGER.debug(
+        "Specific %s response: GetNumberOfDeflections=%s",
+        SWITCH_TYPE_DEFLECTION,
+        deflections_response,
+    )
+
+    if deflections_response["NewNumberOfDeflections"] == 0:
+        _LOGGER.debug("The FRITZ!Box has no %s options", SWITCH_TYPE_DEFLECTION)
+        return []
+
+    deflection_list = get_deflections(fritzbox_tools, service_name)
+    if deflection_list is None:
+        return []
+
+    return [
+        FritzBoxDeflectionSwitch(
+            fritzbox_tools, device_friendly_name, dict_of_deflection
+        )
+        for dict_of_deflection in deflection_list
+    ]
+
+
+def port_entities_list(
+    fritzbox_tools: FritzBoxTools, device_friendly_name: str
+) -> list[FritzBoxPortSwitch]:
+    """Get list of port forwarding entities."""
+
+    _LOGGER.debug("Setting up %s switches", SWITCH_TYPE_PORTFORWARD)
+    entities_list: list = []
+    service_name = "Layer3Forwarding"
+    connection_type = service_call_action(
+        fritzbox_tools, service_name, "1", "GetDefaultConnectionService"
+    )
+    if not connection_type:
+        _LOGGER.debug("The FRITZ!Box has no %s options", SWITCH_TYPE_PORTFORWARD)
+        return []
+
+    # Return NewDefaultConnectionService sample: "1.WANPPPConnection.1"
+    con_type: str = connection_type["NewDefaultConnectionService"][2:][:-2]
+
+    # Query port forwardings and setup a switch for each forward for the current device
+    resp = service_call_action(
+        fritzbox_tools, con_type, "1", "GetPortMappingNumberOfEntries"
+    )
+    if not resp:
+        _LOGGER.debug("The FRITZ!Box has no %s options", SWITCH_TYPE_DEFLECTION)
+        return []
+
+    port_forwards_count: int = resp["NewPortMappingNumberOfEntries"]
+
+    _LOGGER.debug(
+        "Specific %s response: GetPortMappingNumberOfEntries=%s",
+        SWITCH_TYPE_PORTFORWARD,
+        port_forwards_count,
+    )
+
+    for i in range(port_forwards_count):
+
+        portmap = service_call_action(
+            fritzbox_tools,
+            con_type,
+            "1",
+            "GetGenericPortMappingEntry",
+            NewPortMappingIndex=i,
+        )
+
+        if not portmap:
+            _LOGGER.debug("The FRITZ!Box has no %s options", SWITCH_TYPE_DEFLECTION)
+            continue
+
+        _LOGGER.debug(
+            "Specific %s response: GetGenericPortMappingEntry=%s",
+            SWITCH_TYPE_PORTFORWARD,
+            portmap,
+        )
+
+        # We can only handle port forwards of the given device
+        local_ip = get_local_ip()
```

Comment:
Move this outside the loop as we only need to fetch it once

Reply:
Done